implements select()

juantascon opened the pull request:
Initial implementation of select() according to https://github.com/nalundgaard/jsn/issues/10

This implementation uses a few anonymous functions but I'm not sure you guys like this syntax. I could change it to private functions instead or just pass the functions directly to the lists:*

Additionally, I'm not 100% ok with the documentation I added but also I'm not sure I should explain all the input options here or should we just leave it to the user to go to jsn.hrl and figure it out by himself

juantascon committed:
implements select()

nalundgaard commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+    %% This function applies a single condition to a single element
+    ApplyConditionFun = 
+        fun({Path, FilterFun}, E) when is_function(FilterFun) ->
+                erlang:apply(FilterFun, [?MODULE:get(Path, E)]);
```
Please use `get/2` instead of `?MODULE:get/2`.

I think that this will "just work" in newer Erlangs based on the note for `no_auto_import` in the [compile docs](http://erlang.org/doc/man/compile.html):

> As from R14A and forward, the compiler resolves calls without module prefix to local or imported functions before trying with auto-imported BIFs. If the BIF is to be called, use the erlang module prefix in the call, not `{no_auto_import,[{F,A}, ...]}`.

That said, if that doesn't work, I'd prefer you use that directive, i.e.

```erlang
-compile({no_auto_import, [get/2]}).
```

This applies to all the other internal uses of `get/2` as well.

jadeallenx replied:
Would be clearer to use `FilterFun(get(Path, E))` here instead of erlang:apply/2.

nalundgaard commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+                erlang:error(badarg)
+        end,
+    FilteredElements =
+        lists:foldl(fun(Condition, CurrentElements) ->
```
I would guess that it would be considerably more efficient to write a recursive `apply_conditions(E, Conditions)` function here that would apply the conditions to the element in order, bailing on the first one that returns `false`.

Then you can just call 

```erlang
lists:filter(fun(E) -> apply_conditions(E, Conditions) end, Elements).
```

I am guessing that using `foldl` to apply each condition, one at a time, to all elements in the list is probably less efficient in most cases. Certainly, it's a lot harder for me to read.

jadeallenx replied:
By writing your own recursive function you can stop evaluating conditions which do not return true much sooner than if you keep evaluating them as in a foldl.

nalundgaard commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+        end,
+    %% This function applies a single condition to a single element
+    ApplyConditionFun = 
+        fun({Path, FilterFun}, E) when is_function(FilterFun) ->
```
I think this would be a lot more readable as a private `apply_condition/2` function that could be called directly in the `apply_conditions/2` function I suggested previously.

nalundgaard commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+           (FilterFun, E) when is_function(FilterFun) ->
+                erlang:apply(FilterFun, [E]);
+           (_, _) ->
+                erlang:error(badarg)
```
I think it might be more client friendly to do something like this here:

```erlang
throw({error, {invalid_condition, Condition}}
```

I think that a `badarg` this deep into the top-level (client-facing) function is likely to result in a stack trace that is *very* hard to understand.

Maybe if you refactor this to not use anonymous functions, though, that won't be an issue.

jadeallenx replied:
I agree that the error `badarg` here would be pretty opaque, but I'm not necessarily sure it needs to be thrown. What's the behavior in other locations when invalid parameters that need validation are passed into functions?  The behavior here should be the same.

nalundgaard commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+                                    end, CurrentElements)
+                    end, Elements, ConditionList),
+    %% Converts Filtered Elements by applying output specifications
+    OutputFun =
```
This needs to be an internal function, not a case statement, because it needs to be recursively referenced.

nalundgaard commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+-spec select(selection(), json_array()) -> json_array().
+select(OutputSpec, Elements) when is_list(Elements) ->
+    select(OutputSpec, [], Elements);
+select(_OutpuSpec, _Elements) ->
```
s/_OutpuSpec/_OutputSpec

nalundgaard commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+                fun(E) -> ?MODULE:get(Path, E, Default) end;
+            _ when is_list(OutputSpec) ->
+                fun(E) ->
+                    lists:map(fun({value, Path}) -> ?MODULE:get(Path, E);
```
lists:map/2 is right here, I think, but you should probably leverage recursion here to reuse the code for matching `{value, Path}` and `{value, Path, Default}`. You're also missing support for `identity` as an element of a list in an output specification, I think.

nalundgaard commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+                              end, OutputSpec)
+                end;
+            _ ->
+                erlang:error(badarg)
```
Same issue here W.R.T. traceable exceptions. Either this should include the bad value that triggered the input, or it should fail at a function level where the immediate argument that triggered the failure is clear.

jadeallenx replied:
Really agree with this sentiment. Make it easy for users to understand specifically what's bad/invalid when a function dies.

nalundgaard reviewed (changes requested):
Overall it looks mostly right to me, but I have a few points of disagreement with the style here. In general my preference is for simpler, more broken down functions with descriptive names. It makes the code easier to follow and much easier to trace.

nalundgaard commented:
@mrallen1 please add any additional thoughts.

jadeallenx commented:
I haven't read this very carefully yet but in general I am pretty strongly 👎 on anonymous functions in higher order functions unless they're trivial.  I'll take a closer look at this tonight.

jadeallenx commented:
Also, before I sound like a total jerkface - thank you for this code! It's a great start. 💖

nalundgaard commented:
+1 to @mrallen1's comment: this is definitely an excellent start. Thanks very much for doing this. Your unit tests prove that it works as discussed.

jadeallenx commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+           ({Path, Value}, E) ->
+                ?MODULE:get(Path, E) == Value;
+           (FilterFun, E) when is_function(FilterFun) ->
+                erlang:apply(FilterFun, [E]);
```
Suggest `FilterFun(E)` here again, for clarity and less verbosity.

jadeallenx reviewed (changes requested):
This is a great start - looking forward to seeing your revisions.

nalundgaard commented on src/jsn.erl:
```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
+                erlang:apply(FilterFun, [?MODULE:get(Path, E)]);
+           ({Path, Value}, E) ->
+                ?MODULE:get(Path, E) == Value;
+           (FilterFun, E) when is_function(FilterFun) ->
```
One other thing here: use

```erlang
when is_function(FilterFun, 1)
```

Since you know it's supposed to be a 1-arity function, and you're calling it with 1 argument, you should guard that.

nalundgaard commented on include/jsn.hrl:
```diff
@@ -70,6 +70,18 @@
 -type path_element() ::  binary() | json_array_index().
 -type path_elements() :: [ path_element() ].
 
+%% Select conditions
```
I think this should probably say `Select Types`

nalundgaard commented on include/jsn.hrl:
```diff
@@ -70,6 +70,18 @@
 
+%% Select conditions
+%%
+%% Select conditions are passed to the select/2 function for filtering
```
This should probably comment on the type for condition(s).

nalundgaard commented on src/jsn.erl:
```diff
@@ -23,6 +23,8 @@
     path_elements/1
 ]).
+%% select API
+-export([select/2, select/3]).
```
Picky function placement detail: I don't think that this set of functions needs its own category. I think that these functions should come after `find/N` in the `lookup/storage API` section above. Can you also relocate the implementation (of the public functions) to be immediately after the `find/N` function implementations as well?

nalundgaard commented:
@juantascon One more thing I think this should have: a reference in the README. I think it would fall naturally in the [extract data from objects](https://github.com/nalundgaard/jsn#get23-get_list23-and-find34---extract-data-from-objects) section. I think just a brief summary and an integration into the end of the example (put an example or two against the `[User1, User2]` list).

If you prefer, I can make that a follow up PR—just let me know.

juantascon committed:
addresses comments

juantascon commented:
Hey guys, I've added a commit where I addressed most of your comments so please review again.

I changed the input types only a little bit so they are more symmetric and I split the code into private funs as you guys suggested and I think it's more clear and easier to read now.

There's only 2 things left: 
1. regarding erlang:error(): not exactly sure what to return here please see the terms I'm using and if you guys have better suggestions please share
2. readme: will edit it once we've reached a consensus on the implementation

nalundgaard commented on src/jsn.erl:
```diff
@@ -656,6 +686,38 @@ from_proplist(X, _Options) -> X.
+apply_condition([], _Element) ->
+    true;
+apply_condition(_, _) ->
+    erlang:error(invalid_condition).
```
So, I am of two minds on this. On the one hand, as is, it might as well be a `badarg`:

```erlang
> try jsn:select([{value, <<"username">>}], [fuuu], [jsn:new()]) catch C:E -> T = erlang:get_stacktrace(), io:format("Exception ~p:~p, trace: ~p~n", [C,E,T]) end.
Exception error:invalid_condition, trace: [{jsn,apply_condition,2,
                                            [{file,
                                              "/Users/nal/src/personal/erlang/jsn/_build/default/lib/jsn/src/jsn.erl"},
                                             {line,706}]},
                                           {jsn,apply_condition,2,
                                            [{file,
                                              "/Users/nal/src/personal/erlang/jsn/_build/default/lib/jsn/src/jsn.erl"},
                                             {line,697}]},
                                           {jsn,'-select/3-fun-0-',3,
                                            [{file,
                                              "/Users/nal/src/personal/erlang/jsn/_build/default/lib/jsn/src/jsn.erl"},
                                             {line,195}]},
                                           {lists,filtermap,2,
                                            [{file,"lists.erl"},{line,1317}]},
                                           {erl_eval,do_apply,6,
                                            [{file,"erl_eval.erl"},
                                             {line,674}]},
                                           {erl_eval,try_clauses,8,
                                            [{file,"erl_eval.erl"},
                                             {line,904}]},
                                           {shell,exprs,7,
                                            [{file,"shell.erl"},{line,686}]},
                                           {shell,eval_exprs,7,
                                            [{file,"shell.erl"},{line,641}]}]
ok
```

I guess the `invalid_condition` makes it obvious that the problem is a condition, but not which one. I feel like it should be either a bare `badarg` exception (the trace will make it obvious where it is), or, if we use a custom error, it should include the broken arg, i.e.

```erlang
apply_condition(C, _) ->
    erlang:error({invalid_condition, C}).
```

nalundgaard replied:
@mrallen1 curious on your thoughts here.

jadeallenx replied:
I would prefer the error report the precise `invalid_condition`. The stack trace is nice and all, but precision is better, imo.

nalundgaard commented on src/jsn.erl:
```diff
@@ -656,6 +686,38 @@ from_proplist(X, _Options) -> X.
+apply_selection(Selections, Element) when is_list(Selections) ->
+    [apply_selection(Selection, Element) || Selection <- Selections];
+apply_selection(_, _) ->
+    erlang:error(invalid_output_spec).
```
Same issue here, more or less (badarg vs. custom error), but also, what is an `output_spec`? Shouldn't that say `selection` if we're using a custom error? I don't think `output_spec` term was used anywhere else in the code.

jadeallenx replied:
Agree this error can be better stated as an `invalid_selection`

nalundgaard commented on src/jsn.erl:
```diff
@@ -656,6 +686,38 @@ from_proplist(X, _Options) -> X.
+apply_selection({value, Path, Default}, Element) ->
+    get(Path, Element, Default);
+apply_selection(Selections, Element) when is_list(Selections) ->
+    [apply_selection(Selection, Element) || Selection <- Selections];
```
I think it would be more elegant to remove this function clause and explicitly implement an `apply_selections/2` function like:

```erlang
apply_selections(Selections, Element) when is_list(Selections) ->
    [apply_selection(Selection, Element) || Selection <- Selections];
apply_selections(Selection, Element) ->
    apply_selection(Selection, Element).
```

The advantage being, this will not work with selection nesting, e.g. a bizarrely crafted input like

```erlang
[[{value, Path}], [identity], [[{value, Path, Default}, [],[]]]]
```

A construction like that will be more true to the input spec.

nalundgaard commented on src/jsn.erl:
```diff
@@ -656,6 +686,38 @@ from_proplist(X, _Options) -> X.
+    get(Path, Element) == Value;
+apply_condition(ConditionFun, Element) when is_function(ConditionFun, 1) ->
+    ConditionFun(Element);
+apply_condition([Condition|Rest], Element) ->
```
I would also use `apply_conditions/2` here, to prevent incidentally supporting weird nesting of conditions. The second clause of `apply_conditions/2` can handle the case for passing through a singular condition to the `apply_condition/2` function, but I think it would be better from both a style and correctness perspective to avoid overloading `apply_condition/2` to support singular and plural conditions.

nalundgaard commented on src/jsn.erl:
```diff
@@ -172,6 +173,35 @@ find(Path, Subpath, SearchTerm, Object) ->
+    end, Elements);
+select(_Selection, _Conditions, _Elements) ->
+    erlang:error(badarg).
+
```
Can you add one more newline here for consistent function spacing?

nalundgaard reviewed (changes requested):
Much better, @juantascon! It is very nice to read, and has a nice flow to it. Couple of minor comments here. Once that is addressed and we decide the path to follow for bad input exceptions, I think it's ready to go.

jadeallenx commented on test/jsn_tests.erl:
```diff
@@ -499,6 +499,32 @@ path_elements_test_() ->
      ?_assertError(badarg, jsn:path_elements(0)),
      ?_assertError(badarg, jsn:path_elements([10, <<"foo">>]))].
 
+select_test_() ->
```
Good start on the tests for this functionality. Please make sure you test the error conditions in your tests, too.

jadeallenx reviewed (changes requested):
Looking pretty solid. Just a couple more things to sort.

juantascon committed:
addresses comments #2

juantascon commented:
2nd round of comments addressed, @mrallen1 @nalundgaard please review again

nalundgaard commented on src/jsn.erl:
```diff
@@ -656,6 +687,46 @@ from_proplist(X, _Options) -> X.
+apply_condition(ConditionFun, Element) when is_function(ConditionFun, 1) ->
+    ConditionFun(Element);
+apply_condition(_, _) ->
+    erlang:error(invalid_condition).
```
can you make this: 

```erlang
apply_condition(Condition, Element) ->
    erlang:error(badarg, [Condition, Element]).
```

It seems like adding the `Args` as the second argument to `erlang:error(...)` is the key to getting a nice error output that is easy to understant. When I do that, I get output like this:

```erlang
jsn:select([{value, <<"username">>}], [fuuu], [jsn:new()]).
** exception error: bad argument
     in function  jsn:apply_condition/2
        called as jsn:apply_condition(fuuu,[])
     in call from jsn:apply_conditions/2 (jsn/_build/default/lib/jsn/src/jsn.erl, line 692)
     in call from jsn:'-select/3-fun-0-'/3 (jsn/_build/default/lib/jsn/src/jsn.erl, line 195)
     in call from lists:filtermap/2 (lists.erl, line 1317)
```

nalundgaard commented on src/jsn.erl:
```diff
@@ -172,6 +173,36 @@ find(Path, Subpath, SearchTerm, Object) ->
+%%------------------------------------------------------------------------------
+%% @doc Transforms a list of Elements according to a given Selectionification
+%%------------------------------------------------------------------------------
+select(Selection, Elements) when is_list(Elements) ->
```
I think you can take the guard (and second clause) off here actually, and let this all fall down to `select/3`, personally.

nalundgaard commented on src/jsn.erl:
```diff
@@ -172,6 +173,36 @@ find(Path, Subpath, SearchTerm, Object) ->
+        end
+    end, Elements);
+select(_Selection, _Conditions, _Elements) ->
+    erlang:error(badarg).
```
I suggest to make this

```erlang
select(Selection, Elements) ->
    erlang:error(badarg, [Selection, Elements]).
```

nalundgaard commented on src/jsn.erl:
```diff
@@ -656,6 +687,46 @@ from_proplist(X, _Options) -> X.
+apply_selection({value, Path, Default}, Element) ->
+    get(Path, Element, Default);
+apply_selection(_, _) ->
+    erlang:error(invalid_selection).
```
I suggest to make this

```erlang
apply_selection(Selection, Element) ->
    erlang:error(badarg, [Selection, Element]).
```

nalundgaard reviewed (changes requested):
OK, *last round*, I promise! Really sorry.

See my comments below. I am thinking we should just use `badarg` everywhere, and use `erlang:error(badarg, Args)` in order to get the exception processing to record the error correctly in a way that clearly points to the offending call.

Otherwise, I think this is ready to go.

nalundgaard commented:
@juantascon I opened up #14 to adjust my current errors in the library to be more client-friendly as well. :)

juantascon committed:
fixes error handling

juantascon commented:
@nalundgaard I agree those tracebacks look better, please review again

jadeallenx commented on test/jsn_tests.erl:
```diff
@@ -499,6 +499,36 @@ path_elements_test_() ->
+                   jsn:select({value, <<"username">>, <<"No Name">>},
+                              [{<<"username">>, undefined}],
+                              Objects)),
+     ?_assertError(badarg,
```
👍

jadeallenx reviewed (approved):
This looks great. Thanks for taking all the feedback in stride!

nalundgaard commented:
@juantascon Thanks again! 🦄

juantascon commented:
thank you guys, I actually learned a lot from this
